### PR TITLE
Record account suspend/silence time and keep track of domain blocks

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Currently, undoing an instance block will either unblock *every* blocked person on that instance or none (depending on the moderator's choice), because an instance block translates to an individual block for every user of that instance.

This change records the time an user was blocked, and ties the unblock part to blocks performed *exactly* at the same time as the instance block.

~~This is a work in progress because we have to figure out what to do with instances that were blocked before that PR. The code in the PR will *not* unblock them, which is probably not what we want to do.~~ The proposed code will currently unblock anyone that matches the severity of the domain block and who was individually blocked *before* the migration was run.

~~Also need to add tests and make sure the interface is clear.~~

### CODE REVIEW

**Comment by ThibG** [comment]
Updated. The unblock screen won't ask whether you want to unblock *all* existing accounts of that instance anymore, and will instead give the count of accounts that would actually be unblocked.

The accounts affected are those whose block date matches that of the domain block or are `nil`. That is, those who were definitely blocked through the domain block as well as those who were blocked before the feature was merged. There's a few things we could do slightly differently in that regard, but not much.

**Review by Gargron — COMMENTED** [review]
We could remove the booleans

**Comment by ThibG** [comment]
We definitely could, I just wanted to avoid a costly and error-prone migration script.



Le 1 mai 2019 17:55:22 GMT+02:00, Eugen Rochko <notifications@github.com> a écrit :
>Gargron commented on this pull request.
>
>We could remove the booleans
>
>
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/tootsuite/mastodon/pull/10660#pullrequestreview-232629983

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by nightpool** [comment]
I think the idea here is good but removing the boolean columns now is going to save us a ton of headaches down the road wrt our data model.

If you really really don't want to remove the booleans, then we should encapsulate any logic that refers to the booleans or the dates into the account model so that we only have to care about the legacy data model in one place.

**Comment by ThibG** [comment]
I have changed the migration script to set the `silenced_at` and `suspended_at` fields of existing accounts to the date of the corresponding domain block if there is one, and to the current time otherwise.
I have changed the code to not use the `suspended` and `silenced` fields anymore and updated the tests accordingly.
Finally, I have added a post-deployment migration script to remove those columns from the database.

I have **not** tried the migration yet.

**Review by nightpool — CHANGES_REQUESTED** [review]
First pass. I think we can narrow down the amount of files touched by this change, especially spec files as noted below. We should make sure we're using the `suspend!` and `silence!` methods in specs going forward.

File: spec/controllers/application_controller_spec.rb
Comment (ThibG) [reply spec/controllers/application_controller_spec.rb]: I've tried to do that but I'm not quite sure how to do that, or if it is even possible

File: app/models/account.rb
Comment (nightpool) [inline app/models/account.rb]: prefer `.present?` to `!thing.nil?`

File: app/models/account.rb
Comment (nightpool) [inline app/models/account.rb]: `date ||= Time.now`

File: app/models/account.rb
Comment (nightpool) [inline app/models/account.rb]: .present per above

File: app/models/account.rb
Comment (nightpool) [inline app/models/account.rb]: ||= per above

File: app/services/resolve_account_service.rb
Comment (nightpool) [inline app/services/resolve_account_service.rb]: I don't agree with this whitespace change but it's also not a huge deal. this line is semantically different from the lines that follow, so they shouldn't be grouped together

**Review by nightpool — CHANGES_REQUESTED** [review]
now that we can undo domain blocks precisely, is there any argument for leaving in the ability to undo them and NOT unsilence/unsuspend affected users? it seems like an extra level of complexity that doesn't help anyone

**Comment by ThibG** [comment]
Removed the `retroactive` option as suggested, making *all* unblocks/unsilences retroactive.

This removes the checkbox but not the text:

image

File: spec/services/block_domain_service_spec.rb
Comment (nightpool) [inline spec/services/block_domain_service_spec.rb]: both `suspended` and `silenced_at`?

**Review by nightpool — APPROVED** [review]
LGTM % a minor question about the tests

**Comment by JamesABDL** [comment]
I think this has broken the explore page:

```
method=GET path=/explore format=html controller=DirectoriesController action=index status=500 error='ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column accounts.suspended does not exist
```

**Comment by Gargron** [comment]
Need to tootctl cache clear

**Pull request merged**
